Richer path templating and parsing

Floppy opened the pull request:
Templates can now include literal text, making it all a lot more flexible. Parsing also works using the same template. Settings have been rationalised and combined or renamed where appropriate.

<img width="1350" alt="Screenshot 2023-03-08 at 23 34 14" src="https://user-images.githubusercontent.com/3565/223878513-edc47a6c-f703-4c13-aa52-d9e92c61bcaf.png">

ksuquix reviewed (commented):
I think we need a flag on whether or not to parameterize model name / tags / etc on output.

ksuquix commented on app/models/concerns/path_parser.rb:
```diff
@@ -1,63 +1,46 @@
+  return {} if components.nil?
+  components.merge({
+    tags: components[:tags]&.split("/")&.compact_blank,
+    model_id: nil # discard ID, never gonna use it in parsing
```
This *could* be used as a basis to figure out when someone did local file changes.  That might help avoid duplication when someone adjusts stuff manually.

Floppy replied:
true, if the ID exists, we could use it to match if the folder path changes.

Floppy replied:
one for the backlog I think

ksuquix replied:
agreed.

Floppy commented:
Restored the option I removed for auto-tagging based on folder name, but I've made it false by default now and cleaned up the naming. I will add the parameterize option tomorrow - you're right that we need it, at the moment there's an inconsistent approach depending on which component it is.

ksuquix commented on app/controllers/settings_controller.rb:
```diff
@@ -21,9 +21,10 @@ def update
-      SiteSettings.model_path_suffix_model_id = params[:model_tags][:model_path_suffix_model_id]
+    end
+    if current_user.admin? && params[:folders]
+      SiteSettings.model_path_template = params[:folders][:model_path_template]
```
I suggest we force this to have a leading / or not.   Right now it could be either way, which at the very least ends with odd-looking results in the dialog box for organizing files.

ksuquix replied:
Upon further inspection... if you have a leading slash, it will always think your files are not organized... so strip any leading slash here.

Floppy replied:
done!

Floppy commented:
`SiteSettings.safe_folder_names` added 👍🏻
